prow/owners/owners.go

```diff
@@ -0,0 +1,477 @@
+	fullName := fmt.Sprintf("%s/%s", org, repo)
+	sha, err := c.ghc.GetRef(org, repo, "heads/master")
+	if err != nil {
+		return nil, fmt.Errorf("failed to get current SHA for %s: %v", fullName, err)
+	}
+
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	entry, ok := c.cache[fullName]
+	if !ok || entry.sha != sha {
+		// entry is non-existant or stale.
+		gitRepo, err := c.git.Clone(fullName)
+		if err != nil {
+			return nil, fmt.Errorf("failed to clone %s: %v", fullName, err)
+		}
+		defer gitRepo.Clean()
+
+		entry.aliases = loadAliasesFrom(gitRepo.Dir, log)
+		entry.sha = sha
+		c.cache[fullName] = entry
+	}
+
+	return entry.aliases, nil
+}
+
+// LoadRepoOwners returns an up-to-date RepoOwners struct for the specified repo.
+// Note: The returned *RepoOwners should be treated as read only.
+func (c *Client) LoadRepoOwners(org, repo string) (*RepoOwners, error) {
+	log := c.Logger.WithFields(logrus.Fields{"org": org, "repo": repo})
+
+	fullName := fmt.Sprintf("%s/%s", org, repo)
+	mdYaml := c.mdYAMLEnabled(org, repo)
+	sha, err := c.ghc.GetRef(org, repo, "heads/master")
+	if err != nil {
+		return nil, fmt.Errorf("failed to get current SHA for %s: %v", fullName, err)
+	}
+
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	entry, ok := c.cache[fullName]
+	if !ok || entry.sha != sha || entry.owners == nil || entry.owners.enableMDYAML != mdYaml {
+		gitRepo, err := c.git.Clone(fullName)
+		if err != nil {
+			return nil, fmt.Errorf("failed to clone %s: %v", fullName, err)
+		}
+		defer gitRepo.Clean()
+
+		if entry.aliases == nil || entry.sha != sha {
+			// aliases must be loaded
+			entry.aliases = loadAliasesFrom(gitRepo.Dir, log)
+		}
+		entry.owners, err = loadOwnersFrom(gitRepo.Dir, mdYaml, entry.aliases, log)
+		if err != nil {
+			return nil, fmt.Errorf("failed to load RepoOwners for %s: %v", fullName, err)
+		}
+		entry.sha = sha
+		c.cache[fullName] = entry
+	}
+
+	var owners *RepoOwners
+	// Filter collaborators. We must filter the RepoOwners struct even if it came from the cache
+	// because the list of collaborators could have changed without the git SHA changing.
+	collaborators, err := c.ghc.ListCollaborators(org, repo)
+	if err != nil {
+		log.WithError(err).Errorf("Failed to list collaborators while loading RepoOwners. Skipping collaborator filtering.")
+		owners = entry.owners
+	} else {
+		owners = entry.owners.filterCollaborators(collaborators)
+	}
+	return owners, nil
+}
+
+func mdYAMLEnabledFromConfig(ca *config.Agent) func(org, repo string) bool {
+	return func(org, repo string) bool {
+		enabledRepos := ca.Config().Owners.MDYAMLRepos
+		full := fmt.Sprintf("%s/%s", org, repo)
+		for _, elem := range enabledRepos {
+			if elem == org || elem == full {
+				return true
+			}
+		}
+		return false
+	}
+}
+
+func (a RepoAliases) ExpandAlias(alias string) sets.String {
+	if a == nil {
+		return nil
+	}
+	return a[github.NormLogin(alias)]
+}
+
+func (a RepoAliases) ExpandAliases(logins sets.String) sets.String {
+	if a == nil {
```

Comment:
It's unlike that a will be nil once this method is called.

Reply:
`LoadRepoAliases` can return a nil alias map without returning an error. This is to allow the case where there is no OWNERS_ALIASES file and to distinguish it from the case where the file exists, but is empty (map would be non nil, but still empty in that case).
I added a check to see if the map is nil in the `ExpandAlias` function so that the result of `LoadRepoAliases` can always be safely used if no error is returned. `ExpandAliases` also contains this check to avoid looping over the logins if there is no chance of expansion.